Allow Webroot Certificate Creation

simonspa opened the pull request:
This PR supersedes #43 by @rfaircloth-splunk and adds support for the `webroot` generation method as well as requesting HSTS or running in test mode.

Fixes #41 support deployhook (webroot only)
Fixes #42 support for test CA
Fixes #39 Support for webroot

kierenevans commented on defaults/main.yml:
```diff
@@ -36,3 +47,6 @@ certbot_keep_updated: true
 certbot_dir: /opt/certbot
+
+# Restart this service after successfull certificate creation:
+certbot_deployhook: "service {{certbot_create_standalone_stop_services }} restart"
```
This doesn't work with multiple services (or even one service where it prints `service [u'nginx'] restart` instead.
Try this instead:
```suggestion
certbot_deployhook: "service {{ certbot_create_standalone_stop_services | join(' restart && service') }} restart"
```

simonspa commented:
You're right @kierenevans thanks for the fix!

MicroJoe commented:
Thanks for this PR, I am currently using it (rebased it on master) until not merged.

One little issue I have had is that the `/etc/letsencrypt/` directory does not seem to exist right after installing the `certbot` package on Debian 10 Buster, probably because `certbot.service` was not started right after installation (but it should have been, by whom, this role or Debian?).

```
TASK [geerlingguy.certbot : Check if certificate already exists.] **********************************************************************************************
ok: [***]

TASK [geerlingguy.certbot : Create deploy hook] ****************************************************************************************************************
fatal: [***]: FAILED! => {"changed": false, "checksum": "7fd17a05d6d10769b284884b86040525a5894358", "msg": "Destination directory /etc/letsencrypt/renewal-hooks/deploy does not exist"}
```

Manually starting `certbot.service` lead to the creation of the directories, which let the role go fine after.

MicroJoe commented on defaults/main.yml:
```diff
@@ -6,21 +6,32 @@ certbot_auto_renew_hour: "3"
+  --noninteractive --agree-tos
   --email {{ cert_item.email | default(certbot_admin_email) }}
+  {{ '--webroot-path ' if certbot_create_method == 'webroot'  else '' }}
+  {{ cert_item.webroot if certbot_create_method == 'webroot'  else '' }}
```
`cert_item.webroot | default(certbot_webroot)` could be useful in order to avoid to repeat `webroot` for each domain if it is common to all of them (in my case: `/var/www/letsencrypt`, not sure if this could be a good default value).

simonspa replied:
Very good idea!

MicroJoe commented on tasks/create-cert-webroot.yml:
```diff
@@ -0,0 +1,22 @@
+
+- name: Create webroot directory if it doesn't exist yet
+  file:
+    path: "{{ cert_item.webroot }}"
```
Same comment, see https://github.com/MicroJoe/ansible-role-certbot/commit/4a114225885b03ab8c1e865c7409f49953c62396

simonspa commented:
Hi @MicroJoe, Thanks for the suggestion, very good idea. I implemented it and it seems to work for me - care to test?

I also added checks to ensure the config directory actually exists - doesn't hurt probably.

MicroJoe commented:
Thanks, it looks great!
I am currently provisioning a new cloud machine to test that the role works on clean machine now with the config directory verification.

MicroJoe commented:
After testing, I think your last commit is not fixing the issue I encounter.
However, here is what fixed it for me: verify that the hook directory exist before copying it.

```diff
diff --git a/tasks/create-cert-webroot.yml b/tasks/create-cert-webroot.yml
index 5d0ac6d..63466c8 100644
--- a/tasks/create-cert-webroot.yml
+++ b/tasks/create-cert-webroot.yml
@@ -4,6 +4,12 @@
     path: /etc/letsencrypt/live/{{ cert_item.domains | first }}/cert.pem
   register: letsencrypt_cert
 
+- name: Ensure deploy hook directory exists
+  file:
+    path: /etc/letsencrypt/renewal-hooks/deploy
+    state: directory
+    mode: 0755
+
 - name: Create deploy hook
   copy:
     content: "{{ certbot_deployhook }}"
```

It should also be added to `create-cert-standalone.yml` for feature parity, but you get the idea.
As for your last commit, I am not sure if it is useful as this change will supersede your verification, so maybe you can adapt it. The less tasks the faster deployment. ;)

simonspa commented:
Great, thanks for testing - I think I misread the error log you posted earlier...

I removed the unnecessary check again and added your patch. It's not necessary to add anything for standalone creation, there no deployment hook is written.

MicroJoe commented:
Thanks, I have rebased my fork's `webroot` branch to match yours. I have nothing to add to this PR, I find it useful and hope it can be merged.

@geerlingguy any thoughts?

MicroJoe commented:
After playing with this feature a little bit, I have noticed that the renewal-hook installation will occur once per domain, while there is nothing depending on the domains in the renewal-hook creation. This is slowing down the certificate generation loop as it is executed each time.

In https://github.com/MicroJoe/ansible-role-certbot/commit/c841d79d93e96bc7bf5467ded0dd16d1e6490a0e I patch your PR to not do the hook installation in the loop but outside of it. It was done in a hurry, but it can be reworked to go nice with this PR's content.

simonspa commented:
Thanks, @MicroJoe - that was indeed unnecessary. I cherry-picked your commit onto this branch!

MicroJoe commented:
When using this patch on a new machine I have had to change the hook creation to happen before the very first certificate is issued. This fix is visible in this commit: https://github.com/MicroJoe/ansible-role-certbot/commit/2346cd52a262bf2fa89419f29abba0e8ea95cdb7

simonspa commented:
Thank you very much for reporting, @MicroJoe - I have cherry-picked your commit and added it to this branch.

simonspa commented:
@geerlingguy any chance for you to review this and get it in? It is already used by a fair amount of people and has been reviewed by two independent parties. I believe that the functionality this adds to your role is worth the few minutes of review.

simonspa commented:
It's inactive since it's just sitting here and waiting for a merge. I still think it's a feature worth having.

geerlingguy commented:
I still like the feature too, and if you can rebase Lint should pass.

simonspa commented:
Hi @geerlingguy 

sure, no problem, I rebased - but the Linter job fails, as it does on `master`. I have fixed a few things (fixing the line length in your github workflow, installing ansible in addition to ansible-lint` but I don't feel comfortable venturing deeper into fixing the ansible-lint errors in this merge request here - they are unrelated to the code discussed here.

Please check and let me know if I should do anything else.

/Simon

geerlingguy commented:
Oh shoot... I accidentally committed the lint fixes on the wrong branch. I'll get that sorted momentarily.

geerlingguy commented:
All right, please try rebasing again.

simonspa commented:
Done, and linter passes :tada:

simonspa commented:
@geerlingguy I implemented all requested changes - any plan to merge this?

karlmdavis commented on defaults/main.yml:
```diff
@@ -37,3 +52,6 @@ certbot_keep_updated: true
 certbot_dir: /opt/certbot
+
+# Restart this service after successfull certificate creation:
+certbot_deployhook: "service {{ certbot_create_standalone_stop_services | join(' restart && service ') }} restart"
```
@simonspa Might I suggest this minor improvement to this excellent PR?

```suggestion
certbot_deployhook: |                                                                                                                                                                                     
  {% for service in certbot_create_standalone_stop_services %}
  if systemctl list-unit-files | grep -q "^{{ service }}.service"; then
    systemctl restart {{ service }}
  fi
  {% endfor %}
```

I often have a chicken-egg problem where the services I want to create certs for don't/can't exist until after certbot has run. Adding these `if`s avoids that problem by only trying to restart the services if they exist.

simonspa replied:
Thanks, LGTM!

geerlingguy commented on defaults/main.yml:
```diff
@@ -6,21 +6,36 @@ certbot_auto_renew_hour: "3"
+  --noninteractive --agree-tos
   --email {{ cert_item.email | default(certbot_admin_email) }}
+  {{ '--webroot-path ' if certbot_create_method == 'webroot'  else '' }}
+  {{ cert_item.webroot | default(certbot_webroot) if certbot_create_method == 'webroot'  else '' }}
```
```suggestion
  {{ cert_item.webroot | default(certbot_webroot) if certbot_create_method == 'webroot'  else '' }}
```

  {{ cert_item.webroot | default(certbot_webroot) if certbot_create_method == 'webroot' else '' }}

geerlingguy commented on defaults/main.yml:
```diff
@@ -37,3 +52,11 @@ certbot_keep_updated: true
 # Where to put Certbot when installing from source.
 certbot_dir: /opt/certbot
+
+# Restart this service after successfull certificate creation:
```
```suggestion
# Restart this service after successful certificate creation:
```

geerlingguy commented on .github/workflows/release.yml:
```diff
@@ -35,4 +35,8 @@ jobs:
 
       - name: Trigger a new import on Galaxy.
-        run: ansible-galaxy role import --api-key ${{ secrets.GALAXY_API_KEY }} $(echo ${{ github.repository }} | cut -d/ -f1) $(echo ${{ github.repository }} | cut -d/ -f2)
+        run: >-
```
Can you revert this change, only because it's unrelated and is standardized across all 60+ of my Galaxy roles.

simonspa replied:
Done. That was the linter though...

hkcomori commented:
In master, arguments have been added to certbot_create_command.
In webroot, I have errors that these directories are not found when first installation.

https://github.com/geerlingguy/ansible-role-certbot/pull/80/files#diff-00781efe962bc7f1510921c6d0bf5340fccdd1436a7507f76e766a1876a1fb03

simonspa commented:
> In master, arguments have been added to certbot_create_command. In webroot, I have errors that these directories are not found when first installation.
> 
> https://github.com/geerlingguy/ansible-role-certbot/pull/80/files#diff-00781efe962bc7f1510921c6d0bf5340fccdd1436a7507f76e766a1876a1fb03

Thanks for the ping, I have rebased this branch on top of latest `master`. @geerlingguy All your comments have been addressed some five months ago, would you mind merging this? :)

/S

geerlingguy commented:
Released in tag 5.0.0—please test and make sure things are working for you!

simonspa commented:
Thanks, working flawlessly over here!
